accepted/2024/observable_collection.md

```diff
@@ -0,0 +1,173 @@
+`ObservableCollection<T>` raises events for individual items and crash when the
+event args represent more than one item, specifically WPF.
+
+From a UI standpoint, bulk operations are desirable because they allow for UI
+controls to be more performant when the data changes. Other UI frameworks that
+support view models, such as Avalonia, Uno, and MAUI have expressed interest in
+seeing this change. In fact, the [tracking issue][issue] is eight years old and
+has over four hundred comments, with several folks expressing frustration that
+this issue hasn't been addressed yet.
+
+The goal of this document is to explore ways to address this request.
+
+## Requirements
+
+### Goals
+
+* Enable bulk operations for `Add`, `Remove`, `Insert`, and `Replace` that
+  result in a single notification for view models.
+
+### Non-Goals
+
+* Support bulk notifications for discontinuous ranges in the collection
+
+## Stakeholders and Reviewers
+
+* Platform components
+    - Libraries team
+    - WinForms team
+    - WPF team
+    - MAUI team
+* External UI platforms
+    - Avalonia
+    - Uno
+
+## Design
+
+### API Changes
+
+We don't need to make any changes to `INotifyCollectionChanged` as
+`NotifyCollectionChangedEventArgs` can already represent ranges:
+
+```C#
+public partial class NotifyCollectionChangedEventArgs : EventArgs
+{
+    // Constructors omitted
+    public NotifyCollectionChangedAction Action { get; }
+    public IList? NewItems { get; }
+    public IList? OldItems { get; }
+    public int NewStartingIndex { get; }
+    public int OldStartingIndex { get; }
+}
+```
+
+> [!NOTE]
+> Please note that the API shape only supports batched notifications if the
+> affected range is contiguous, for example, inserting multiple items or
+> removing a contiguous range. It doesn't, for example, support removing
+> multiple discontiguous items.
+
+`ObservableCollection<T>` needs to expose new APIs for performing bulk
+operations that will result in bulk notifications:
+
+```C#
+namespace System.Collections.ObjectModel;
+
+public class ObservableCollection<T>
+{
+    public void AddRange(IEnumerable<T> collection);
+    public void InsertRange(int index, IEnumerable<T> collection);
+    public void RemoveRange(int index, int count);
+    public void ReplaceRange(int index, int count, IEnumerable<T> collection);
+
+    protected virtual void InsertItemsRange(int index, IEnumerable<T> collection);
+    protected virtual void RemoveItemsRange(int index, int count);
+    protected virtual void ReplaceItemsRange(int index, int count, IEnumerable<T> collection);
+}
+```
+
+## How it breaks consumers
+
+`ObservableCollection<T>` exists since .NET Framework 3.0, which was shipped in
+2006, 18 years ago. While the eventing structure in principle supports raising
+events for more than one item, the collection was never capable of raising such
+events. As a result, consumers of `ObservableCollection<T>`, specifically WPF,
```

Comment:
Ultimately, this is what we're paying for is assumptions made long ago... I hope we'll be able to overcome this for the benefit of a much broader ecosystem that exists now impacted by this.